// ngx-countdown/esm2020/countdown.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { CountdownComponent } from './countdown.component';
import { CountdownTimer } from './countdown.timer';
import * as i0 from "@angular/core";
export class CountdownModule {
}
CountdownModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: CountdownModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
CountdownModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: CountdownModule, declarations: [CountdownComponent], imports: [CommonModule], exports: [CountdownComponent] });
CountdownModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: CountdownModule, providers: [CountdownTimer], imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: CountdownModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    providers: [CountdownTimer],
                    declarations: [CountdownComponent],
                    exports: [CountdownComponent],
                }]
        }] });
